1.14 Payment validation fix

rust-le opened the pull request:
| Q               | A
|-----------------|-----
| Branch?         | 1.14
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes
| License         | MIT

This fix blocks the ability to place an order if the payment method is blocked for a given channel, but is still active

rust-le committed:
Add Resolver to OrderPaymentMethodEligibilityValidator

rust-le committed:
Moved checker to Component/Payment

NoResponseMate commented on src/Sylius/Component/Payment/Checker/OrderPaymentMethodEligibilityCheckerInterface.php:
```diff
@@ -0,0 +1,21 @@
+
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+
+interface OrderPaymentMethodEligibilityCheckerInterface
```
Move this interface and the implementation to Core.
The Payment component is a separate context from orders.

rust-le replied:
done

NoResponseMate commented on src/Sylius/Component/Payment/Checker/OrderPaymentMethodEligibilityChecker.php:
```diff
@@ -0,0 +1,36 @@
+    {
+    }
+
+    public function isEligible(?PaymentMethodInterface $paymentMethod): bool
+    {
+        try {
+            $channel = $this->channelContext->getChannel();
+        } catch (ChannelNotFoundException) {
+            return false;
+        }
+
+        return $paymentMethod->isEnabled() && $paymentMethod->hasChannel($channel);
+    }
```
Although it's simple logic, it should be covered with a spec.

rust-le replied:
done

rust-le committed:
Moved checker to Core

rust-le committed:
Add PHPUnit and Behat tests

rust-le committed:
Fix in behat test

rust-le committed:
Behat scenario @api not work as expected

rust-le committed:
Fixed PaymentContext method

rust-le committed:
Add PerChannel checker, behat and unit tests

# Conflicts:
#	tests/Checker/Eligibility/OrderPaymentMethodPerChannelEligibilityCheckerTest.php

rust-le committed:
Add PerChannel checker, behat and unit tests

rust-le committed:
Removed errors from OrderPaymentMethodEligibilityValidatorSpec

rust-le committed:
Add methods to cover all variants OrderPaymentMethodEligibilityValidatorSpec

TheMilek commented on src/Sylius/Bundle/CoreBundle/Validator/Constraints/OrderPaymentMethodEligibilityValidator.php:
```diff
@@ -14,12 +14,21 @@
 
 final class OrderPaymentMethodEligibilityValidator extends ConstraintValidator
 {
+    public function __construct(
+        private readonly OrderPaymentMethodEligibilityCheckerInterface $checker,
+        private readonly OrderPaymentMethodPerChannelEligibilityCheckerInterface $perChannelChecker,
+    ) {
+    }
```
I was more thinking of creating separate Validator  to avoid addition of constructor to this one 🤔
So like creating for example `OrderPaymentMethodEnabledInChannelValidator`

TheMilek commented on features/checkout/order_payment_method_eligibility_check.feature:
```diff
@@ -0,0 +1,51 @@
+Feature: Preventing placing an order with a disabled payment method when in checkout complete step
+    In order to have my order shipped without issues
+    As a Customer
+    I want to be prevented from placing an order with a disabled shipping method
```
```suggestion
    I want to be prevented from placing an order with a payment method not available in my channel
```

rust-le replied:
done

TheMilek commented on features/checkout/order_payment_method_eligibility_check.feature:
```diff
@@ -0,0 +1,51 @@
+@checkout
+Feature: Preventing placing an order with a disabled payment method when in checkout complete step
+    In order to have my order shipped without issues
+    As a Customer
```
```suggestion
    As a Visitor
```

rust-le replied:
ok

TheMilek commented on features/checkout/order_payment_method_eligibility_check.feature:
```diff
@@ -0,0 +1,51 @@
+@checkout
+Feature: Preventing placing an order with a disabled payment method when in checkout complete step
```
```suggestion
Feature: Preventing placing an order with a payment method not available in channel
```

TheMilek commented on features/checkout/order_payment_method_eligibility_check.feature:
```diff
@@ -0,0 +1,51 @@
+        And I should not see the thank you page
+
+    @ui @api
+    Scenario: Being able placing an order with a payment method that's has been not disabled globally and/or for Channel after completing the shipping method choice step
```
```suggestion
    Scenario: Being able to place an order with enabled payment method in my channel
```
here and in others I think

rust-le replied:
ok

TheMilek commented on features/checkout/order_payment_method_eligibility_check.feature:
```diff
@@ -0,0 +1,51 @@
+        And I should not see the thank you page
+
+    @ui @api
+    Scenario: Being prevented from placing an order with a payment method that's has been disabled globally and for Channel after completing the shipping method choice step
```
```suggestion
    Scenario: Being prevented from placing an order with a payment method that is disabled and not available in my channel
```

rust-le replied:
ok

TheMilek commented on src/Sylius/Behat/Context/Setup/PaymentContext.php:
```diff
@@ -184,6 +184,27 @@ public function theLatestOrderHasAuthorizedPayment(OrderInterface $order, string
+
+        $this->paymentMethodManager->flush();
+
+        $this->sharedStorage->set('disabled_payment_method', $paymentMethod);
```
```suggestion
```
I think its not used anywhere if I see correctly 🤔

rust-le replied:
done

TheMilek commented on src/Sylius/Behat/Context/Setup/PaymentContext.php:
```diff
@@ -184,6 +184,27 @@ public function theLatestOrderHasAuthorizedPayment(OrderInterface $order, string
     }
 
+    /**
+     * @Given the store has disabled :paymentMethodName payment method in Channel :channel
+     */
+    public function theStoreHasDisabledPaymentMethodInChannel($paymentMethodName, ChannelInterface $channel)
+    {
+        $paymentMethods = $this->paymentMethodRepository->findByName($paymentMethodName, 'en_US');
+
+        if (empty($paymentMethods)) {
+            throw new \InvalidArgumentException(sprintf('Payment method "%s" does not exist.', $paymentMethodName));
+        }
+
+        /** @var PaymentMethodInterface $paymentMethod */
+        $paymentMethod = $paymentMethods[0];
```
```suggestion
     * @Given the store has disabled :paymentMethod payment method in Channel :channel
     */
    public function theStoreHasDisabledPaymentMethodInChannel(PaymentMethodInterface $paymentMethod, ChannelInterface $channel)
    {
```

I think you can do here the same like it's done here https://github.com/Sylius/Sylius/blob/1.14/src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php#L515-L524

TheMilek replied:
Here is how this step i linked is used somewhere else https://github.com/Sylius/Sylius/blob/1.14/features/shipping/applying_shipping_fee/applying_correct_shipping_fee_with_product_taxes_on_order.feature#L37

rust-le replied:
ok

rust-le committed:
Add changes suggested in CR

rust-le committed:
Fixed minor errors, all tests pass

rust-le committed:
Restred OrderPaymentMethodEligibilityValidator, cleanup

TheMilek commented on features/checkout/order_payment_method_eligibility_check.feature:
The name of the file should be `preventing_placing_an_order_with_a_payment_method_not_available_in_channel.feature`

rust-le replied:
fixed

TheMilek commented on src/Sylius/Bundle/CoreBundle/Resources/config/validation/Order.xml:
```diff
@@ -28,6 +28,12 @@
         </constraint>
+        <constraint name="Sylius\Bundle\CoreBundle\Validator\Constraints\OrderPaymentMethodPerChannelEligibility">
+            <option name="groups">
+                <value>sylius</value>
```
```suggestion

```

rust-le replied:
done

TheMilek commented on src/Sylius/Bundle/CoreBundle/Validator/Constraints/OrderPaymentMethodPerChannelEligibility.php:
```diff
@@ -0,0 +1,31 @@
+
+use Symfony\Component\Validator\Constraint;
+
+final class OrderPaymentMethodPerChannelEligibility extends Constraint
```
```suggestion
final class OrderPaymentMethodEnabledInChannel extends Constraint
```
Maybe something like that here in other places 🤔

rust-le replied:
or rather OrderPaymentMethodEnabledInChannelEligibility ? And in corresponding places to

rust-le replied:
Changed as suggested

TheMilek commented on config/services.yaml:
```diff
@@ -1 +0,0 @@
-
```
why?

rust-le replied:
this file was temporary not empty

rust-le committed:
Fixes from CR

NoResponseMate commented on features/checkout/preventing_placing_an_order_with_a_payment_method_not_availale_in_channel.feature:
```diff
@@ -0,0 +1,51 @@
+        And the store allows paying "Bank Transfer"
+        And I am a logged in customer
+
+    @ui @api
```
```suggestion
    @api @ui
```
Same below

NoResponseMate commented on src/Sylius/Behat/Context/Setup/PaymentContext.php:
```diff
@@ -184,6 +184,17 @@ public function theLatestOrderHasAuthorizedPayment(OrderInterface $order, string
     }
 
+    /**
+     * @Given the store has disabled :paymentMethod payment method in Channel :channel
```
```suggestion
     * @Given the :paymentMethod payment method has been disabled in :channel
```

NoResponseMate commented on src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderPaymentMethodEligibilityValidator.php:
```diff
@@ -36,14 +37,19 @@ public function validate(mixed $value, Constraint $constraint): void
         /** @var PaymentInterface $payment */
         foreach ($order->getPayments() as $payment) {
-            if (!$payment->getMethod()->isEnabled()) {
+            $paymentMethod = $payment->getMethod();
+            if ($paymentMethod instanceof PaymentMethodInterface && (!$paymentMethod->isEnabled() || !$paymentMethod->hasChannel($channel))) {
```
```suggestion
            /** @var PaymentMethodInterface $paymentMethod */
            $paymentMethod = $payment->getMethod();
            if (!$paymentMethod->isEnabled() || !$paymentMethod->hasChannel($channel)) {
```
We're in API context, it's already at the very least a Core payment method

NoResponseMate commented on src/Sylius/Component/Core/Payment/Checker/OrderPaymentMethodEnabledInChannelEligibilityChecker.php:
```diff
@@ -0,0 +1,36 @@
+    {
+    }
+
+    public function isEligible(?PaymentMethodInterface $paymentMethod): bool
+    {
+        try {
+            $channel = $this->channelContext->getChannel();
+        } catch (ChannelNotFoundException) {
+            return false;
+        }
+
+        return $paymentMethod->hasChannel($channel);
+    }
```
This whole service seems unnecessary, also adds a stateful layer

TheMilek commented:
Closing as it is fixed in https://github.com/Sylius/Sylius/pull/18941
